### PR TITLE
Feature/softmax softplus activation

### PR DESCRIPTION
**Pull request by d4n-hewitt**
Adding the softmax and the softplus activation functions

### CODE REVIEW

File: tests/unit/activations/test_activations.py [hunk 1]
```diff
@@ -56,20 +58,40 @@ def test_from_config_method_returns_instance_of_activation():
 
 
 @pytest.mark.parametrize(
```
Comment (pytergeist) [inline tests/unit/activations/test_activations.py:75]: I think this has now gotten to the point where it's worth splitting these into two tests - parametrised and non parametrised activation fns

File: tests/unit/activations/test_activations.py [hunk 2]
```diff
@@ -38,6 +39,7 @@ def test_name_method_returns_snake_case_class_name():
     [
         ("relu", ReLU),
         ("leaky_relu", LeakyReLU),
+        ("softmax", Softmax),
```
Comment (pytergeist) [inline tests/unit/activations/test_activations.py:56]: shouldn't softplus be in here as well

File: tests/unit/activations/test_activations.py [hunk 3]
```diff
@@ -23,6 +23,7 @@ def build(self, input_shape: Optional[Tuple[int, ...]] = None) -> None:
     [
         ("ReLU", "relu"),
         ("LeakyReLU", "leakyrelu"),
+        ("Softmax", "softmax"),
```
Comment (pytergeist) [inline tests/unit/activations/test_activations.py:39]: softplus

File: nova/src/blocks/activations/activations.py [hunk 4]
```diff
@@ -34,26 +34,43 @@ def __init__(self):
...
 
-class Softplus:
-    def __init__(self):
-        raise NotImplementedError("Softplus activation function is not implemented.")
+    def get_config(self):
+        return {}
+
+
+class Softplus(Block):
+    def __init__(self, beta: float = 1.0, **kwargs):
```
Comment (pytergeist) [inline nova/src/blocks/activations/activations.py:68]: is this 0.1 the commonly used generic param?

**Review by pytergeist — COMMENTED** [review]
few comments

File: nova/src/blocks/activations/activations.py (hunk 4 above)
Comment (d4n-hewitt) [reply nova/src/blocks/activations/activations.py:68]: 1.0 is the default in pytorch: https://docs.pytorch.org/docs/2.12/generated/torch.nn.Softplus.html

File: tests/unit/activations/test_activations.py (hunk 3 above)
Comment (d4n-hewitt) [reply tests/unit/activations/test_activations.py:39]: [f477b92](https://github.com/pytergeist/Nova/pull/34/commits/f477b92ecacbfc469eb939bb71d365f6ea68d6e0)

File: tests/unit/activations/test_activations.py (hunk 2 above)
Comment (d4n-hewitt) [reply tests/unit/activations/test_activations.py:56]: [f477b92](https://github.com/pytergeist/Nova/pull/34/commits/f477b92ecacbfc469eb939bb71d365f6ea68d6e0)

File: tests/unit/activations/test_activations.py (hunk 1 above)
Comment (d4n-hewitt) [reply tests/unit/activations/test_activations.py:75]: [f97ad63](https://github.com/pytergeist/Nova/pull/34/commits/f97ad63656a2329fce57ef024bc9aea4f21ba79e)

File: tests/unit/activations/test_activations.py [hunk 5]
```diff
@@ -58,38 +60,62 @@ def test_from_config_method_returns_instance_of_activation():
...
-            0,
-            0,
         ),  # TODO: Add larger input value test when max is implemented
-        ("softplus", Tensor([2.0, 1.0]), [2.126928011, 1.31326168], 0, 1),
-        ("softplus", Tensor([2.0, 1.0]), [2.6265233, 1.948154], 0, 0.5),
     ],
 )
-def test_call_method(
-    activation_fn, data, expected, negative_slope, inverse_temperature
+def test_non_paramaterised_call_method(activation_fn, data, expected):
```
Comment (pytergeist) [inline tests/unit/activations/test_activations.py]: change this name

File: tests/unit/activations/test_activations.py [hunk 6]
```diff
@@ -58,38 +60,62 @@ def test_from_config_method_returns_instance_of_activation():
...
+
+@pytest.mark.parametrize(
+    "activation_fn, data, expected, negative_slope",
+    [
+        ("leaky_relu", Tensor([1.0, 1.0]), [1, 1], 0.1),
+        ("leaky_relu", Tensor([-1.0, 1.0]), [-0.1, 1], 0.1),
+        ("leaky_relu", Tensor([-1.0, 1.0]), [-0.5, 1], 0.5),
+    ],
+)
+def test_negative_sloped_param_call_method(
```
Comment (pytergeist) [inline tests/unit/activations/test_activations.py]: same here - this one is for a single fn so include leaky_relu in the name

File: tests/unit/activations/test_activations.py [hunk 7]
```diff
@@ -58,38 +60,62 @@ def test_from_config_method_returns_instance_of_activation():
...
+
+
+@pytest.mark.parametrize(
+    "activation_fn, data, expected, inverse_temperature",
+    [
+        ("softplus", Tensor([2.0, 1.0]), [2.126928011, 1.31326168], 1),
+        ("softplus", Tensor([2.0, 1.0]), [2.6265233, 1.948154], 0.5),
+    ],
+)
+def test_inverse_temperature_param_call_method(
```
Comment (pytergeist) [inline tests/unit/activations/test_activations.py]: same here - have softplus in the name

File: tests/unit/activations/test_activations.py (hunk 5 above)
Comment (d4n-hewitt) [reply tests/unit/activations/test_activations.py]: [f96741c](https://github.com/pytergeist/Nova/pull/34/commits/f96741c7558a043e4d0750fb31dc22fc54342870)

File: tests/unit/activations/test_activations.py (hunk 6 above)
Comment (d4n-hewitt) [reply tests/unit/activations/test_activations.py]: [f96741c](https://github.com/pytergeist/Nova/pull/34/commits/f96741c7558a043e4d0750fb31dc22fc54342870)

File: tests/unit/activations/test_activations.py (hunk 7 above)
Comment (d4n-hewitt) [reply tests/unit/activations/test_activations.py]: [f96741c](https://github.com/pytergeist/Nova/pull/34/commits/f96741c7558a043e4d0750fb31dc22fc54342870)

File: tests/unit/activations/test_activations.py [hunk 8]
```diff
@@ -18,6 +18,17 @@ def build(self, input_shape: Optional[Tuple[int, ...]] = None) -> None:
         pass
 
 
+def _util_activation_call_test(activation_fn, data, expected, **kwargs):
```
Comment (pytergeist) [inline tests/unit/activations/test_activations.py]: lets put 'allclose' in the name, and pass rtol/atol as args

Comment (d4n-hewitt) [reply tests/unit/activations/test_activations.py]: [544ded2](https://github.com/pytergeist/Nova/pull/34/commits/544ded2d1367479058570c026e33343836f57164)

**Review by pytergeist — APPROVED** [review]
lgtm
